File: LeetCode/Medium/0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        List<int[]>[] graph = new ArrayList[n];

        for (int i = 0; i < n; i++)
            graph[i] = new ArrayList<>();

        for (int[] f : flights)
            graph[f[0]].add(new int[] { f[1], f[2] });

        int[] stops = new int[n];
        Arrays.fill(stops, Integer.MAX_VALUE);

        PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> o1[1] - o2[1]);
        pq.offer(new int[] { src, 0, 0 });
    
        while (!pq.isEmpty()) {
            int[] curr = pq.poll();

            if (curr[0] == dst) return curr[1];

            if (curr[2] > k || curr[2] >= stops[curr[0]]) continue;
            stops[curr[0]] = curr[2];

            for (int[] next : graph[curr[0]]) {
                pq.offer(new int[] { next[0], curr[1] + next[1], curr[2] + 1 });
            }
        }

        return -1;
    }
}

File: LeetCode/Medium/0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.kt
class Solution {
    fun findCheapestPrice(n: Int, flights: Array<IntArray>, src: Int, dst: Int, k: Int): Int {
        val graph = Array(n) { mutableListOf<IntArray>() }
        for (f in flights) {
            graph[f[0]].add(intArrayOf(f[1], f[2]))
        }

        val stops = IntArray(n) { Int.MAX_VALUE }
        val pq = PriorityQueue<IntArray> { o1, o2 -> o1[1] - o2[1] }
        pq.offer(intArrayOf(src, 0, 0))

        while (pq.isNotEmpty()) {
            val curr = pq.poll()

            if (curr[0] == dst) return curr[1]

            if (curr[2] > k || curr[2] >= stops[curr[0]]) continue
            stops[curr[0]] = curr[2]

            for (next in graph[curr[0]]) {
                pq.offer(intArrayOf(next[0], curr[1] + next[1], curr[2] + 1))
            }
        }

        return -1
    }
}